lib/Sema/CSDiagnostics.cpp

```diff
@@ -2430,6 +2433,101 @@ bool ContextualFailure::tryTypeCoercionFixIt(
   return false;
 }
 
+bool ContextualFailure::tryProtocolConformanceFixIt(
+    InFlightDiagnostic &diagnostic) const {
+  auto innermostTyCtx = getDC()->getInnermostTypeContext();
+  if (!innermostTyCtx)
+    return false;
+
+  auto nominal = innermostTyCtx->getSelfNominalTypeDecl();
+  if (!nominal)
+    return false;
+
+  // We need to get rid of optionals and parens as it's not relevant when
+  // printing the diagnostic and the fix-it.
+  auto unwrappedToType =
+      ToType->lookThroughAllOptionalTypes()->getWithoutParens();
+
+  // If the protocol requires a class & we don't have one (maybe the context
+  // is a struct), then bail out instead of offering a broken fix-it later on.
+  auto requiresClass = false;
+  if (unwrappedToType->isExistentialType()) {
+    if (auto protocolTy = unwrappedToType->getAs<ProtocolType>()) {
+      requiresClass = protocolTy->requiresClass();
+    } else if (auto compositionTy =
+                   unwrappedToType->getAs<ProtocolCompositionType>()) {
+      requiresClass = compositionTy->requiresClass();
+    }
+  }
+  if (requiresClass && !FromType->is<ClassType>()) {
+    return false;
+  }
+
+  // We can only offer a fix-it if we're assigning to a protocol type and
+  // the type we're assigning is the same as the innermost type context.
+  bool shouldOfferFixIt = nominal->getSelfTypeInContext()->isEqual(FromType) &&
+                          unwrappedToType->isExistentialType();
+  if (!shouldOfferFixIt)
+    return false;
+
+  diagnostic.flush();
+
+  // Let's build a list of protocols that the contextual type does not
+  // conform to. We will start by first checking if we have a protocol
+  // composition type and add all the individual types that the context
+  // does not conform to.
+  SmallVector<std::string, 8> missingProtoTypeStrings;
+  if (auto compositionTy = unwrappedToType->getAs<ProtocolCompositionType>()) {
```

Comment:
There is an easier way to do this without having to check for `ProtocolCompositionType` explicitly, take a look at `matchExistentialTypes`, you can get a layout from existential and ask it about protocols.

Reply:
Thanks, I have made the change!

Reply:
Done! The existential layout tip was really helpful and makes the code simpler!